File: frontend/src/services/routing.ts
import { get, post } from '@/lib/api';

import { rankVehiclesForLoad } from './assignment';

import type {
  AssignPlanResult,
  AssignmentScore,
  GeoPoint,
  GeoSearchResult,
  RouteLeg,
  RouteResult,
  Stop,
  Vehicle,
  VehicleType,
} from '@/types';

export interface WaypointPayload {
  name: string;
  location: GeoPoint;
}

export interface OptimizePayload {
  start: WaypointPayload;
  stops: WaypointPayload[];
  destination: WaypointPayload;
  profile: 'driving-hgv' | 'driving-car';
  reorderStops: boolean;
  weightKg: number;
  vehicleCostPerKm?: number;
  mileageKmpl?: number;
  capacityKg?: number;
}

function toWaypoint(stop: { name: string; location: GeoPoint }): WaypointPayload {
  return { name: stop.name, location: stop.location };
}

export function searchPlaces(text: string, country = 'IN'): Promise<GeoSearchResult[]> {
  return get<GeoSearchResult[]>('/v1/geo/search', { text, country });
}

export function reverseGeocode(lat: number, lng: number): Promise<string> {
  return get<{ label: string }>('/v1/geo/reverse', { lat, lng }).then((d) => d.label);
}

export async function optimizeRoute(payload: OptimizePayload): Promise<RouteResult> {
  const data = await post<RouteResult>('/v1/routes/optimize', {
    start: toWaypoint(payload.start),
    stops: payload.stops.map(toWaypoint),
    destination: toWaypoint(payload.destination),
    profile: payload.profile,
    weight_kg: payload.weightKg,
    reorder_stops: payload.reorderStops,
    vehicle: {
      operating_cost_per_km: payload.vehicleCostPerKm ?? 48,
      mileage_kmpl: payload.mileageKmpl ?? 4.2,
      capacity_kg: payload.capacityKg ?? 16000,
    },
  });
  return { ...data, steps: [], provider: 'ors' };
}

export async function planAssignment(
  start: Stop,
  stops: Stop[],
  destination: Stop,
  weightKg: number,
  profile: 'driving-hgv' | 'driving-car',
  vehicles: Vehicle[],
): Promise<AssignPlanResult> {
  return post<AssignPlanResult>('/v1/assign/plan', {
    start: toWaypoint(start),
    stops: stops.map(toWaypoint),
    destination: toWaypoint(destination),
    weight_kg: weightKg,
    profile,
    vehicles: vehicles.map((v) => ({
      id: v.id,
      name: v.name,
      plate_number: v.plateNumber,
      vehicle_type: v.type,
      capacity_kg: v.capacityKg,
      operating_cost_per_km: v.operatingCostPerKm,
      mileage_kmpl: v.mileageKmpl,
      fuel_type: v.fuelType,
      health_score: v.healthScore,
      available: v.available,
    })),
  });
}

export function profileForType(type: VehicleType): 'driving-hgv' | 'driving-car' {
  return type === 'heavy-truck' || type === 'mini-truck' ? 'driving-hgv' : 'driving-car';
}

export type { AssignmentScore };

const EARTH_RADIUS_KM = 6371.0088;

function toRad(deg: number): number {
  return (deg * Math.PI) / 180;
}

export function haversineKm(a: GeoPoint, b: GeoPoint): number {
  const { sin, cos, asin, sqrt } = Math;
  const dLat = toRad(b.lat - a.lat);
  const dLng = toRad(b.lng - a.lng);
  const h =
    sin(dLat / 2) ** 2 + cos(toRad(a.lat)) * cos(toRad(b.lat)) * sin(dLng / 2) ** 2;
  return 2 * EARTH_RADIUS_KM * asin(sqrt(h));
}

export function straightLineGeometry(a: GeoPoint, b: GeoPoint, steps = 24): GeoPoint[] {
  const out: GeoPoint[] = [];
  for (let i = 0; i <= steps; i++) {
    out.push({
      lat: a.lat + ((b.lat - a.lat) * i) / steps,
      lng: a.lng + ((b.lng - a.lng) * i) / steps,
    });
  }
  return out;
}

/**
 * Client-only planner used when the backend (or its ORS key) is unreachable.
 * Straight-line geometry + haversine distance + fleet scoring, so route
 * optimization always produces an answer even fully offline.
 */
export function buildOfflineAssignment(
  start: Stop,
  stops: Stop[],
  destination: Stop,
  weightKg: number,
  profile: 'driving-hgv' | 'driving-car',
  vehicles: Vehicle[],
): AssignPlanResult {
  const ordered = [start, ...stops, destination];
  const avgSpeed = profile === 'driving-hgv' ? 48 : 40;
  const costPerKm = profile === 'driving-hgv' ? 48 : 20;
  const mileageKmpl = profile === 'driving-hgv' ? 4.2 : 12;

  const legs: RouteLeg[] = [];
  const geometry: GeoPoint[] = [];
  let distanceKm = 0;
  let durationMin = 0;

  for (let i = 0; i < ordered.length - 1; i++) {
    const a = ordered[i];
    const b = ordered[i + 1];
    const dist = haversineKm(a.location, b.location);
    const durMin = (dist / avgSpeed) * 60;
    const seg = straightLineGeometry(a.location, b.location);
    distanceKm += dist;
    durationMin += durMin;
    legs.push({ fromName: a.name, toName: b.name, distanceKm: dist, durationMin: durMin });
    if (geometry.length > 0) geometry.push(...seg.slice(1));
    else geometry.push(...seg);
  }

  const fuel = distanceKm / mileageKmpl;
  const route: RouteResult = {
    distanceKm: Math.round(distanceKm * 100) / 100,
    durationMin: Math.round(durationMin),
    geometry,
    steps: [],
    fuelEstimateL: Math.round(fuel),
    costEstimate: Math.round(distanceKm * costPerKm),
    co2EstimateKg: Math.round(fuel * 2.68),
    provider: 'local',
    profile,
    congestionPct: 0,
    legs,
    orderedWaypoints: ordered,
  };

  const ranked = rankVehiclesForLoad({ vehicles, weightKg, distanceKm });
  return { route, recommended: ranked[0] ?? null, ranked };
}

export interface ResilientPlan extends AssignPlanResult {
  /** True when the backend (or its ORS dependency) could not be reached. */
  offline: boolean;
  /** The real, surfaceable reason for the fallback, when known. */
  offlineReason?: string;
}

export async function planAssignmentResilient(
  start: Stop,
  stops: Stop[],
  destination: Stop,
  weightKg: number,
  profile: 'driving-hgv' | 'driving-car',
  vehicles: Vehicle[],
): Promise<ResilientPlan> {
  try {
    const plan = await planAssignment(start, stops, destination, weightKg, profile, vehicles);
    return { ...plan, offline: false };
  } catch (err) {
    const reason =
      (err instanceof Error ? err.message : String(err)) || 'Backend request failed';
    console.error('[routing] Backend assignment failed — real reason:', reason, err);
    const plan = buildOfflineAssignment(start, stops, destination, weightKg, profile, vehicles);
    return { ...plan, offline: true, offlineReason: reason };
  }
}

File: backend/app/schemas/routing.py
"""Pydantic schemas for routing, geocoding and vehicle assignment."""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, Field

RouteProfile = Literal["driving-car", "driving-hgv", "cycling-regular"]


class GeoPoint(BaseModel):
    lat: float = Field(ge=-90, le=90)
    lng: float = Field(ge=-180, le=180)


class Waypoint(BaseModel):
    name: str = ""
    location: GeoPoint


class VehicleParams(BaseModel):
    """Vehicle economics used for cost estimation (optional overrides)."""

    operating_cost_per_km: float = 48.0
    mileage_kmpl: float = 4.2
    capacity_kg: float = 16000.0
    health_score: int = 100
    fuel_type: str = "diesel"


class OptimizeRequest(BaseModel):
    start: Waypoint
    stops: list[Waypoint] = Field(default_factory=list)
    destination: Waypoint
    profile: RouteProfile = "driving-hgv"
    vehicle: VehicleParams | None = None
    weight_kg: float = 0.0
    reorder_stops: bool = True


class RouteLeg(BaseModel):
    from_name: str
    to_name: str
    distance_km: float
    duration_min: float


class GeoSearchResult(BaseModel):
    name: str
    label: str
    region: str
    country: str
    longitude: float
    latitude: float


class OptimizeResult(BaseModel):
    provider: Literal["ors"] = "ors"
    profile: RouteProfile
    distance_km: float
    duration_min: float
    congestion_pct: float
    fuel_estimate_l: float
    cost_estimate: float
    co2_estimate_kg: float
    geometry: list[GeoPoint]
    legs: list[RouteLeg]
    ordered_waypoints: list[Waypoint]


class VehicleCandidate(BaseModel):
    id: str
    name: str
    plate_number: str = ""
    vehicle_type: str = "heavy-truck"
    capacity_kg: float
    operating_cost_per_km: float
    mileage_kmpl: float
    fuel_type: str = "diesel"
    health_score: int = 100
    available: bool = True


class AssignPlanRequest(BaseModel):
    start: Waypoint
    stops: list[Waypoint] = Field(default_factory=list)
    destination: Waypoint
    weight_kg: float = 0.0
    profile: RouteProfile = "driving-hgv"
    vehicles: list[VehicleCandidate] = Field(default_factory=list)


class AssignmentScore(BaseModel):
    vehicle_id: str
    vehicle_name: str
    plate_number: str
    vehicle_type: str
    availability: bool
    capacity_ok: bool
    score: float
    cost_estimate: float
    co2_estimate_kg: float
    eta_min: float
    reasons: list[str]


class AssignPlanResult(BaseModel):
    route: OptimizeResult
    recommended: AssignmentScore | None
    ranked: list[AssignmentScore]